mitxgraders/comparers/linear_comparer.py

```diff
@@ -0,0 +1,169 @@
+
+    Configuration
+    =============
+    The first four configuration keys determine the amount of partial credit
+    given for a specific type of linear relationship. If set to None, the
+    relationship is not checked.
+        equals (None | number): defaults to 1.0
+        proportional (None | number): defaults to 0.5
+        offset (None | number): defaults to None
+        linear (None | number): defaults to None
+
+    The remaining configuration keys specify a feedback message to be given
+    in each case:
+        equals_msg (str): defaults to ''
+        proportional_msg (str): defaults to 'The submitted answer differs from
+            an expected answer by a constant factor.'
+        offset_msg (str): defaults to ''
+        linear_msg (str): defaults to ''
+
+    NOTE:
+    """
+
+    schema_config = Schema({
+        Required('equals', default=1.0): Any(None, Range(0, 1)),
+        Required('proportional', default=0.5): Any(None, Range(0, 1)),
+        Required('offset', default=None): Any(None, Range(0, 1)),
+        Required('linear', default=None): Any(None, Range(0, 1)),
+        Required('equals_msg', default=''): str,
+        Required('proportional_msg', default=(
+            'The submitted answer differs from an expected answer by a '
+            'constant factor.'
+        )): str,
+        Required('offset_msg', default=''): str,
+        Required('linear_msg', default=''): str,
+    })
+
+    def __init__(self, config=None, **kwargs):
+        super(LinearComparer, self).__init__(config, **kwargs)
+        all_modes = ('equals', 'proportional', 'offset', 'linear')
+        self.modes = tuple(mode for mode in all_modes if self.config[mode] is not None)
+
+    error_calculators = {
+        'equals': get_equals_fit_error,
+        'proportional': get_proportional_fit_error,
+        'offset': get_offset_fit_error,
+        'linear': get_linear_fit_error,
+    }
+
+    def __call__(self, comparer_params_evals, student_evals, utils):
+        student_eval_norm = np.linalg.norm(student_evals)
+
+        # Validate student input shape...only needed for MatrixGrader
+        try:
+            utils.validate_shape(student_evals[0], comparer_params_evals[0][0].shape)
+        except AttributeError:
+            pass # not called by MatrixGrader
+
+        if is_nearly_zero(student_eval_norm, utils.tolerance, reference=comparer_params_evals):
+            return False
+
+        # Get the result for each mode
+        student = np.array(student_evals).flatten()
+        expected = np.array(comparer_params_evals).flatten()
+        errors = [self.error_calculators[mode](student, expected) for mode in self.modes]
+        results = [
+            {'grade_decimal': self.config[mode], 'msg': self.config[mode+'_msg']}
+            if is_nearly_zero(error, utils.tolerance, reference=student_eval_norm)
+            else
+            {'grade_decimal': 0, 'msg': ''}
+            for mode, error in zip(self.modes, errors)
```

Comment:
This is a bit unusual - the comparer is setting the `msg`. It's also possible for the answer to have a `msg` setting associated with it. I'm just noting this here because we should probably make sure that any message set here isn't overwritten? I'll read on...

Reply:
Ok, looks like the `msg` from a comparer is ignored if the comparer evaluates true. See comment below.